feat / add connector for BitMart

srirambandi opened the pull request:
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:

- Related GitHub issue: #3741
- Add a connector for [BitMart Exchange](https://www.bitmart.com/)
- Research / Investigation issue: #3742

RC-13 commented:
Hi @srirambandi, here are some of the issues that we've encountered during testing.
- When connecting the BitMart API key, it appears that as long as you enter the correct API key and enter an incorrect secret and memo, the key still connects/gets accepted.
- There is a delay with updating the available  balance after opening order, this may result in inconsistent order creation 
![bitmart-available-balance-delay](https://user-images.githubusercontent.com/57189990/128665074-0221afe1-6972-401f-bd48-caf1196e9e55.gif)
- Best price and order book are getting out of sync
![bitmart-best-price1](https://user-images.githubusercontent.com/57189990/128665133-b7390ed9-b7e1-46df-bdd3-85521a2f92b9.png)
![bitmart-order-book-out-of-sync](https://user-images.githubusercontent.com/57189990/128665137-3a9764d2-9187-44d4-b8fc-3e336b4e38ff.gif)
- Also the price precision doesn't match the exchange
![bitmart-precision](https://user-images.githubusercontent.com/57189990/128665258-2a4261cd-c54a-4c2f-8cbf-98500e4f138f.png)
![bitmart-best-price1](https://user-images.githubusercontent.com/57189990/128665263-a76bae6c-750f-467f-a568-d2108ebbc4a3.png)

rxlxrxsx commented:
- There are instances that the bot fails to capture/record filled trades. In this screenshot you'll see that the order was filled however upon checking on history the trade wasn't recorded, the bottom nav bar also shows that there were 0 trades. The said order is also not found on the csv and sqlite file. 2 out of 6 trades were not recorded based on testing.

<img width="1422" alt="Screen Shot 2021-08-09 at 3 24 32 PM" src="https://user-images.githubusercontent.com/64377055/128675177-1b413f8f-2d58-4c4a-b723-7fd8c008bd3c.png">
<img width="1253" alt="Screen Shot 2021-08-09 at 3 25 04 PM" src="https://user-images.githubusercontent.com/64377055/128675184-b230170c-4758-4fb3-b501-8d17c58a423a.png">

[bitmart_filled.zip](https://github.com/CoinAlpha/hummingbot/files/6952840/bitmart_filled.zip)

srirambandi commented:
Hi @RC-13, thanks for the feedback.

> * When connecting the BitMart API key, it appears that as long as you enter the correct API key and enter an incorrect secret and memo, the key still connects/gets accepted.

````connect <exchange_name>```` checks balances api on the exchange. BitMart uses only API Key for getting balances of user. Related [link](https://developer-pro.bitmart.com/en/part2/permission.html). check permissions for ````/spot/v1/wallet````  

> * There is a delay with updating the available  balance after opening order, this may result in inconsistent order creation

BitMart doesn't have websockets for balance updates yet. That could be the reason for the delayed updates, as we are relying on wallet API in _status_polling_function in exchange file. We need to communicate with BitMart to provide websockets for balances for real-time updates.

> * Best price and order book are getting out of sync

> * Also the price precision doesn't match the exchange

Fixed these. Please check.

srirambandi commented:
Hi @rxlxrxsx 

> * There are instances that the bot fails to capture/record filled trades. In this screenshot you'll see that the order was filled however upon checking on history the trade wasn't recorded, the bottom nav bar also shows that there were 0 trades. The said order is also not found on the csv and sqlite file. 2 out of 6 trades were not recorded based on testing.

BitMart [trades API](https://developer-pro.bitmart.com/en/spot/order/trades.html) has detail_id as trade id for each filled trade, where as websocket [trade channel](https://developer-pro.bitmart.com/en/ws/spot_ws/user_trade.html) doesn't provide detail_id for the trade. So, I went a little complex by calling [_update_trade_status function](https://github.com/srirambandi/hummingbot/blob/feat/bitmart-connector/hummingbot/connector/exchange/bitmart/bitmart_exchange.py#L557) which uses API. I am using trade channel for order status updates, and [stopping order tracking](https://github.com/srirambandi/hummingbot/blob/feat/bitmart-connector/hummingbot/connector/exchange/bitmart/bitmart_exchange.py#L700) based on order status rather than trade fulfillment. 

We can use websockets trade channel and treat the trade timestamp as the id and get things done? Please look into the functions and advise me on how to proceed.

RC-13 commented:
@srirambandi Confirming that issue reported related to precision and order book/best price synchronization has been fixed.
Adding another unusual behavior, when placing an order that is below the exchange minimum requirement, the orders are attempted to be submitted however they will be immediately removed and no cancellation message is given. After the refresh period, it will repeat the same process.
![BitMart-order-below-minimum](https://user-images.githubusercontent.com/57189990/129137313-0f447758-d428-4dd2-99b2-6f489aab7ada.gif)

rxlxrxsx commented:
Adding another issue possibly related to balance updates, I noticed that there are instances that the bot takes time before placing another set of orders.

<img width="1423" alt="Screen Shot 2021-08-12 at 1 12 12 PM" src="https://user-images.githubusercontent.com/64377055/129146118-f7a38138-2a97-4e4b-88c0-8d111dfb77db.png">

Logs and config: [bitmart1.zip](https://github.com/CoinAlpha/hummingbot/files/6973181/bitmart1.zip)

srirambandi commented:
> @srirambandi Confirming that issue reported related to precision and order book/best price synchronization has been fixed.
> Adding another unusual behavior, when placing an order that is below the exchange minimum requirement, the orders are attempted to be submitted however they will be immediately removed and no cancellation message is given. After the refresh period, it will repeat the same process.

Thanks for confirming resolving previous issues. 
Can you please share the config file for the above two strategies, so I can try to reproduce them and resolve them from my end? Thanks.

RC-13 commented:
Hi @srirambandi here's my config, you can easily replicate the issue by overriding the `min_quote_order_amount` under `conf_global.yml` file then intentionally set an order amount that is below the exchange minimum allowable trade size. The behavior seems to differ when running from the cloud, now I'm getting an error. Can we have the error message improve so that users can easily identify why the order is not getting submitted?
![image](https://user-images.githubusercontent.com/57189990/129200684-a0ee9b44-adc3-4c0f-8d2e-e8397b7fb1d2.png)
![image](https://user-images.githubusercontent.com/57189990/129201861-7784991e-8545-4c00-8b86-f22e5f95c6d1.png)

srirambandi commented:
Hi @RC-13, I have made some changes and pushed the fixes. Of the issues mentioned above, I fixed the seemingly urgent one of the trades not being recorded(2nd one).

> 1. The error reporting  has improve, although there is an unusual behavior when bot is running on a local machine vs on the cloud.
> 
> * bot running on local machine shows a logging message when order fails to be submitted
> * bot running on a cloud server didn't show any logging message for an order that fails to be submitted

I will get back to you on this.

> 2. The issue with failing to record some filled orders still persist. In my test, there are 3 filled order however only 2 trades are recorded. Correct number of filled trades are shown in the trade history in the exchange.

I added fixes to this. Please check. Due to the way BitMart gives out the data in REST and WebSocket APIs, I had to bootstrap trade recording functions by adding capabilities to process trades from rest and ws. The compromise for this is to get estimated fee instead of the real traded fee(suing estimate_fee util).

> 3. There are random instances as well that the strategy takes time to start after the order book is initialize. The delay may take 1 to 2 min.

I observed this during the testing. During the 'silent' period, in the ````status --live```` mode, it reads 'Market connectors not ready'. So this must be some issue related to API connection or something. I'll investigate more and also push a fix for this.

So, I pushed fixes for 2nd issue and I am working to solve the other two. Please check the pushed fixes. Thanks!

srirambandi commented:
Hi @RC-13 ,

Pushed a fix to solve the 3rd issue too, where the bot stops for 5 sec to 2 mins sometimes. This was due to lack of real time account balance updates from websockets. So, I changed the poll intervals' lengths to maintain a regular status poll every tick(1 second). 

Please check. Thanks.

RC-13 commented:
@srirambandi Thank you for pushing the fix, confirming that issue reported: 1) unrecorded order/missing order and 2) delay during startup of the strategy have been fixed. We are now running a long term test bot and report back if a new issue is found.

dennisocana commented:
@srirambandi the connector does not meet the required minimum unit test coverage and cannot be merged.

![image](https://user-images.githubusercontent.com/50150287/131689847-df18019c-e6a3-491b-b70e-6c4d746a64c8.png)

You may use the sample unit test cases implemented in NDAX or Bybit Perpetual connectors.

srirambandi commented:
Hi @dennisocana, I have started implementing the unit tests. Will push them as soon as they are available.

RC-13 reviewed (approved):
After running a test bot for more than a week, there are no new issues seen and all filled orders has been tracked. LGTM!

dennisocana commented:
Update from developer: Half of required unit tests are complete and working on the remaining tasks within this week. API throttler update will be submitted today.

srirambandi commented:
Hi @dennisocana, fixed the unit tests. please check and review. 

Thanks!

CrimsonJacket commented on hummingbot/connector/exchange/bitmart/bitmart_api_order_book_data_source.py:
```diff
@@ -0,0 +1,248 @@
+    @classmethod
+    async def get_last_traded_prices(cls, trading_pairs: List[str]) -> Dict[str, float]:
+        result = {}
+        async with aiohttp.ClientSession() as client:
+            async with client.get(f"{CONSTANTS.REST_URL}/{CONSTANTS.GET_LAST_TRADING_PRICES_PATH_URL}", timeout=10) as response:
+                response_json = await response.json()
+                for ticker in response_json["data"]["tickers"]:
+                    t_pair = bitmart_utils.convert_from_exchange_trading_pair(ticker["symbol"])
+                    if t_pair in trading_pairs and ticker["last_price"]:
+                        result[t_pair] = float(ticker["last_price"])
+        return result
```
You are not consuming the `AsyncThrottler` at all here.  Wrap the selection in an async context as you did for `BitmartExchange`

```suggestion
        async with self._throttler.execute_task(CONSTANTS.GET_LAST_TRADING_PRICES_PATH_URL):
            async with aiohttp.ClientSession() as client:
                async with client.get(f"{CONSTANTS.REST_URL}/{CONSTANTS.GET_LAST_TRADING_PRICES_PATH_URL}", timeout=10) as response:
                    response_json = await response.json()
                    for ticker in response_json["data"]["tickers"]:
                        t_pair = bitmart_utils.convert_from_exchange_trading_pair(ticker["symbol"])
                        if t_pair in trading_pairs and ticker["last_price"]:
                            result[t_pair] = float(ticker["last_price"])
        return result
```

srirambandi replied:
Hi @CrimsonJacket, I have added the api throttler encapsulations to all the rest api functions in order book data source file.

However, adding api throttler to ws functions in _order_book_data_source or _user_stream_data_source with exchange specified limits, arise warning of api exceeding limits, whereas leaving them as with current timeout parameters and configurations doesn't raise any errors, works seamlessly. So, haven't extended the api throttler to ws.

Please check the made changes and review.

Thanks!

CrimsonJacket reviewed (changes requested):
It seems you have not consuming the `AsyncThrottler` for the `BitmartAPIOrderBookDataSource`. Since there are websocket limits, you will have to wrap the relevant logic in the async context as well.

CrimsonJacket reviewed (approved):
`AsyncThrottler` integration comment is marked as resolved. PR LGTM.
@srirambandi Just a note, it appears that there is no limits for WebSocket API as seen [here](https://developer-pro.bitmart.com/en/limit/).

dennisocana commented:
Merging into development to unblock the unit test tasks.
